daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -2094,6 +2120,37 @@ private[lf] object SBuiltin {
 
   }
 
+  /** $cacheDisclosedContract[T] :: ContractId T -> CachedContract T -> Unit */
+  private[speedy] final case class SBCacheDisclosedContract(contractId: V.ContractId)
+      extends OnLedgerBuiltin(1) {
+
+    override protected def executeWithLedger(
+        args: util.ArrayList[SValue],
+        machine: Machine,
+        onLedger: OnLedger,
+    ): Control = {
+      val cachedContract = extractCachedContract(machine, args.get(0))
+      val templateId = cachedContract.templateId
+      val optError = for {
```

Comment:
Code to populate the disclosed contract key table

Reply:
Code now accounts for:
- contract key hashing errors
- and disclosure key table errors

Reply:
In this case, it looks like the suggested change might introduce some code duplication. I'll add in an explicit type annotation here to help lighten the cognitive overhead.

Reply:
In general, I totally agree that this type of monadic stacking is something to be avoided.